egraph.rkt

```diff
@@ -45,6 +53,39 @@
 
 ;; Only ever use leaders as keys!
 (struct egraph (cnt leader->iexprs expr->parent) #:mutable)
+(define merge-time 0)
+(define rebuild-time 0)
+
+
+(define (egraph-rebuild eg)
+  (define before-time (current-inexact-milliseconds))
+  (define res
+    (egraph-rebuild-loop eg))
+  (set! rebuild-time (+ rebuild-time (- (current-inexact-milliseconds) before-time)))
+  res)
+
+(define (egraph-rebuild-loop eg)
+  (unless (equal? (egraph-rebuild-once eg) 0)
+    (egraph-rebuild-loop eg)))
+
+(define (egraph-rebuild-once eg)
+  (define new-memo (make-hash))
+  (define to-union empty)
+  ;; loop over all exprs and add them
+  (for ([(leader _iexprs) (egraph-leader->iexprs eg)])
+    (for ([node (cons leader (enode-children leader))])
+      (define expr (update-en-expr (enode-expr node)))
+      (define old-leader
+        (hash-ref new-memo expr #f))
+      (hash-set! new-memo expr leader)
+      (unless (or (equal? old-leader #f) (equal? old-leader leader))
+        (set! to-union (cons (list old-leader leader) to-union)))))
+  
+  (for ([pair to-union])
+    (merge-egraph-nodes! eg (first pair) (second pair) #t)
+    (dedup-vars! (first pair)))
```

Comment:
`merge-egraph-nodes` only dedups vars when you call it with two nodes from the same eclass that are different. This happens sometimes during upwards merging. If we deduped on each upwards merge it would be expensive.